### PR TITLE
[mlir][affine] Add unit tests for `isProjectedPermutation`

### PR DESCRIPTION
**Pull request by banach-space**
The only way to test `isProjectedPermutation` is through unit tests. The
concept of "projected permutations" is tricky to document and these
tests are a good source documentation of the expected/intended
behavoiur. Hence these additional unit tests.

### CODE REVIEW

File: mlir/unittests/IR/AffineMapTest.cpp [hunk 1]
```diff
@@ -21,3 +21,44 @@ TEST(AffineMapTest, inferMapFromAffineExprs) {
   map.replace(replacements);
   EXPECT_EQ(map, map);
 }
+
+TEST(AffineMapTest, isProjectedPermutation) {
+  MLIRContext ctx;
+  OpBuilder b(&ctx);
+
+  // 1. Empty map - a projected permutation.
```
Comment (javedabsar1) [inline mlir/unittests/IR/AffineMapTest.cpp:29]: its a 'projected permutation' test and TRUE, FALSE tells the expected so no need to repeat 'a projected permutation'

File: mlir/unittests/IR/AffineMapTest.cpp [hunk 2]
```diff
@@ -21,3 +21,44 @@ TEST(AffineMapTest, inferMapFromAffineExprs) {
...
+
+  // 6. A constant result that's not a {0} - not a projected permutation.
+  auto one = b.getAffineConstantExpr(1);
+  AffineMap map6 = AffineMap::get(1, 0, {one}, &ctx);
+  EXPECT_FALSE(map6.isProjectedPermutation(/*allowZeroInResults=*/true));
+
+  // 7. Not a dim expression - not a projected permutation.
+  auto d0 = b.getAffineDimExpr(0);
+  auto d1 = b.getAffineDimExpr(1);
+
```
Comment (javedabsar1) [inline mlir/unittests/IR/AffineMapTest.cpp:60]: Maybe add some tougher tests
(d0,d1, d2, d3,d4,d5) ->(d5,d3,d0,d1,d2,d4)
(d0,d1, d2, d3,d4,d5) ->(d5,d3,d0+d1,d2,d4)
(d0,d1, d2, d3,d4,d5) ->(d5,d3,d2,d4)

**Review by javedabsar1 — CHANGES_REQUESTED** [review]
Thanks for this. ProjectedPermutation is used in many places and these are very instructive tests.

File: mlir/unittests/IR/AffineMapTest.cpp (hunk 1 above)
Comment (banach-space) [reply mlir/unittests/IR/AffineMapTest.cpp:29]: Good suggestion, thanks!

File: mlir/unittests/IR/AffineMapTest.cpp (hunk 2 above)
Comment (banach-space) [reply mlir/unittests/IR/AffineMapTest.cpp:60]: Yup, thanks, about to add these.

**Review by javedabsar1 — APPROVED** [review]
Looks great now. Thanks.

**Pull request merged**